py/farm_ng/core/event_service_recorder.py

```diff
@@ -0,0 +1,109 @@
+"""
+# run the event_service which starts a simple test publisher
+python -m farm_ng.core.event_service
+# run the event_recorder which subscribes to the test publisher and records the events to a file
+python -m farm_ng.core.event_service_recorder record config.json --config_name=record_all foo
+# note that the config file has an EventServiceConfig with name "record_all" which subscribes to the test publisher
+
+# then try playing back log file:
+python -m farm_ng.core.events_file_reader playback foo.0000.bin
+"""
+from __future__ import annotations
+
+import argparse
+import asyncio
+from pathlib import Path
+import sys
+
+from farm_ng.core.event_client import EventClient
+from farm_ng.core.event_service_pb2 import (
+    EventServiceConfigList,
+    SubscribeRequest,
+)
+from farm_ng.core.events_file_reader import proto_from_json_file
+from farm_ng.core.events_file_writer import EventsFileWriter
+from farm_ng.core.uri import uri_query_to_dict
+from farm_ng.core.stamp import get_monotonic_now, StampSemantics
+
+
+class EventServiceRecorder:
+    def __init__(self, service_name, config_list: EventServiceConfigList):
+        self.service_name = service_name
+        self.config_list = config_list
+        self.recorder_config = None
+        self.clients = dict()
+        for config in self.config_list.configs:
+            if self.service_name == config.name:
+                self.recorder_config = config
+            if config.port != 0:
+                self.clients[config.name] = EventClient(config)
+
+        assert (
+            self.recorder_config is not None
+        ), f"service {self.service_name} not found in config list {self.config_list}"
+        self.record_queue: asyncio.Queue = asyncio.Queue()
+
+    async def record(
+        self, file_base: str | Path, extension: str = ".bin", max_file_mb: int = 0
+    ):
+        with EventsFileWriter(file_base, extension, max_file_mb) as writer:
+            while True:
+                event, payload = await self.record_queue.get()
+                event.timestamps.append(
+                    get_monotonic_now(semantics=StampSemantics.FILE_WRITE)
+                )
+                writer.write_event_payload(event, payload)
+
+    async def subscribe(self, client: EventClient, subscription: SubscribeRequest):
+        async for event, payload in client.subscribe(subscription, decode=False):
+            await self.record_queue.put((event, payload))
+
+    async def subscribe_and_record(
```

Comment:
Actually we want no await puts, so we don't queue messages, and small queue sizes. We may drop messages but want this to not put back pressure on sender.